pkg/controller/bastion/actuator_delete.go

```
@@ -0,0 +1,142 @@
+	}
+
+	if !deleted {
+		return &ctrlerror.RequeueAfterError{
+			RequeueAfter: 10 * time.Second,
+			Cause:        errors.New("bastion instance is still deleting"),
+		}
+	}
+
+	err = removeSecurityGroup(openstackClientFactory, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove security group: %w", err)
+	}
+	return nil
+}
+
+func removeBastionInstance(logger logr.Logger, openstackClientFactory openstackclient.Factory, opt *Options) error {
+	instances, err := getBastionInstance(openstackClientFactory, opt.BastionInstanceName)
+	if openstackclient.IgnoreNotFoundError(err) != nil {
+		return err
+	}
+
+	if len(instances) == 0 {
+		return nil
+	}
+
+	err = deleteBastionInstance(openstackClientFactory, instances[0].ID)
+	if err != nil {
+		return fmt.Errorf("failed to terminate bastion instance: %w", err)
+	}
+
+	logger.Info("Instance removed", "instance", opt.BastionInstanceName)
+	return nil
+}
+
+func removePublicIPAddress(logger logr.Logger, openstackClientFactory openstackclient.Factory, opt *Options) error {
+	fips, err := getFipByName(openstackClientFactory, opt.BastionInstanceName)
+	if err != nil {
+		return err
+	}
+
+	if len(fips) == 0 {
+		return nil
+	}
+
+	err = deleteFloatingIP(openstackClientFactory, fips[0].ID)
+	if err != nil {
+		return fmt.Errorf("failed to terminate bastion Public IP: %w", err)
+	}
+
+	logger.Info("Public IP removed", "public IP ID", fips[0].ID)
+	return nil
+}
+
+func removeSecurityGroup(openstackClientFactory openstackclient.Factory, opt *Options) error {
+	bastionSecurityGroups, err := getSecurityGroups(openstackClientFactory, opt.SecurityGroup)
+	if err != nil {
+		return err
+	}
+
+	if len(bastionSecurityGroups) == 0 {
+		return nil
+	}
+
+	return deleteSecurityGroup(openstackClientFactory, bastionSecurityGroups[0].ID)
+}
+
+func isInstanceDeleted(openstackClientFactory openstackclient.Factory, opt *Options) (bool, error) {
+	instances, err := getBastionInstance(openstackClientFactory, opt.BastionInstanceName)
```

err is not handled properly. Only the ignore not found is covered.